File: icon.go
// ABOUTME: Icon loading from various sources (file, URL, base64).
// ABOUTME: Handles resolution priority and image scaling.

package main

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"image"
	_ "image/jpeg"
	"image/png"
	"net/http"
	"os"
	"strings"
	"time"
)

// loadIconFromBase64 decodes base64 image data and returns a scaled image.
// Handles both raw base64 and data URI format (data:image/png;base64,...)
func loadIconFromBase64(data string) (image.Image, error) {
	// Strip data URI prefix if present
	if idx := strings.Index(data, ";base64,"); idx != -1 {
		data = data[idx+8:]
	}

	decoded, err := base64.StdEncoding.DecodeString(data)
	if err != nil {
		return nil, fmt.Errorf("base64 decode: %w", err)
	}

	img, _, err := image.Decode(bytes.NewReader(decoded))
	if err != nil {
		return nil, fmt.Errorf("image decode: %w", err)
	}

	return scaleImage(img, iconSize, iconSize), nil
}

// loadIconFromURL fetches an image from a URL and returns a scaled image.
func loadIconFromURL(url string) (image.Image, error) {
	client := &http.Client{
		Timeout: 10 * time.Second,
	}

	resp, err := client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("fetch URL: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, resp.Status)
	}

	img, _, err := image.Decode(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("image decode: %w", err)
	}

	return scaleImage(img, iconSize, iconSize), nil
}

// loadIconFromPath loads an image from a file path and returns a scaled image.
func loadIconFromPath(path string) (image.Image, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	img, _, err := image.Decode(f)
	if err != nil {
		return nil, fmt.Errorf("image decode: %w", err)
	}

	return scaleImage(img, iconSize, iconSize), nil
}

// resolveIcon loads an icon from a notification using priority: iconData > iconHref > iconPath.
// Returns nil, nil if no icon is specified.
func resolveIcon(n Notification) (image.Image, error) {
	if n.IconData != "" {
		return loadIconFromBase64(n.IconData)
	}
	if n.IconHref != "" {
		return loadIconFromURL(n.IconHref)
	}
	if n.IconPath != "" {
		return loadIconFromPath(n.IconPath)
	}
	return nil, nil
}

// encodeImageToBase64 encodes an image to a base64 PNG string.
func encodeImageToBase64(img image.Image) (string, error) {
	var buf bytes.Buffer
	if err := png.Encode(&buf, img); err != nil {
		return "", fmt.Errorf("png encode: %w", err)
	}
	return base64.StdEncoding.EncodeToString(buf.Bytes()), nil
}

File: daemon/src/icon.rs
// Icon loading from various sources: base64 data, local file path, URL.
// Handles decoding, scaling to a standard size, and preparing for GPU upload.

use base64::Engine;
use tracing::warn;

use crate::notification::NotificationPayload;

pub const ICON_SIZE: u32 = 48;

/// Try to load an icon synchronously from the payload (base64 or file path).
/// Returns None if no sync source is available or loading fails.
pub fn resolve_sync(payload: &NotificationPayload) -> Option<image::RgbaImage> {
    if !payload.icon_data.is_empty() {
        match load_from_base64(&payload.icon_data) {
            Ok(img) => return Some(img),
            Err(e) => warn!("Failed to decode icon base64: {}", e),
        }
    }
    if !payload.icon_path.is_empty() {
        match load_from_file(&payload.icon_path) {
            Ok(img) => return Some(img),
            Err(e) => warn!("Failed to load icon from {}: {}", payload.icon_path, e),
        }
    }
    None
}

/// Returns true if the payload has a URL icon source that requires async fetch.
/// Only relevant when base64 and file path are both empty.
pub fn needs_async_fetch(payload: &NotificationPayload) -> bool {
    payload.icon_data.is_empty()
        && payload.icon_path.is_empty()
        && !payload.icon_href.is_empty()
}

/// Fetch an icon from a URL. Call from a tokio task.
pub async fn load_from_url(url: &str) -> anyhow::Result<image::RgbaImage> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()?;
    let bytes = client.get(url).send().await?.bytes().await?;
    let img = image::load_from_memory(&bytes)?;
    Ok(scale_to_icon(img))
}

/// Public entry for center to load stored icon_data.
pub fn load_from_base64_pub(data: &str) -> anyhow::Result<image::RgbaImage> {
    load_from_base64(data)
}

fn load_from_base64(data: &str) -> anyhow::Result<image::RgbaImage> {
    // Strip data URI prefix if present (e.g. "data:image/png;base64,...")
    let b64 = if let Some(pos) = data.find(";base64,") {
        &data[pos + 8..]
    } else {
        data
    };
    let decoded = base64::engine::general_purpose::STANDARD.decode(b64)?;
    let img = image::load_from_memory(&decoded)?;
    Ok(scale_to_icon(img))
}

fn load_from_file(path: &str) -> anyhow::Result<image::RgbaImage> {
    let img = image::open(path)?;
    Ok(scale_to_icon(img))
}

fn scale_to_icon(img: image::DynamicImage) -> image::RgbaImage {
    // Scale down preserving aspect ratio, then center on a transparent canvas
    let scaled = img.resize(ICON_SIZE, ICON_SIZE, image::imageops::FilterType::Lanczos3);
    let sw = scaled.width();
    let sh = scaled.height();

    if sw == ICON_SIZE && sh == ICON_SIZE {
        return scaled.to_rgba8();
    }

    let mut canvas = image::RgbaImage::new(ICON_SIZE, ICON_SIZE);
    let ox = (ICON_SIZE - sw) / 2;
    let oy = (ICON_SIZE - sh) / 2;
    image::imageops::overlay(&mut canvas, &scaled.to_rgba8(), ox as i64, oy as i64);
    canvas
}
